pkgs.formats.vimAtoms: init

S0AndS0 opened the pull request:
The output of `generate` does not prefixed `let ${name} =` because there are other places formatted results could be injected within Vim scripts.

... Plus the verity of prefixed options of `let` alone makes ideas of maintaining such things not spark joy; check `:help let` for details.

This means, until I'm told a better way, users will need to use `.text` from `.generate` result to access formatted output, ex;

```nix
nix-repl> git-nixpkgs = import ./default.nix { }
nix-repl> f = git-nixpkgs.pkgs.formats.vim { }
nix-repl> x = f.generate "test" { foo.bar = "woot"; }
nix-repl> :p x.text
#> {"foo":{"bar":"woot"}}
```

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
  - [x] Tests in [pkgs/pkgs-lib/tests/formats.nix]
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

S0AndS0 committed:
pkgs.formats.vimAtoms: init

The output of `generate` does not prefixed `let ${name} =` because there
are other places formatted results could be injected within Vim scripts.

... Plus the verity of prefixed options of `let` alone makes ideas of
maintaining such things not spark joy; check `:help let` for details.

:warning: Floating point conversion from Nix to Vim script, as well as
Vim's floating point representation, will be a consistent source of
opportunities for frustration.

Co-authored-by: h7x4 <h7x4@nani.wtf>

Stunkymonkey commented on pkgs/pkgs-lib/tests/formats.nix:
```diff
@@ -921,6 +921,46 @@ runBuildTests {
+    input = {
+      attrs.foo = "woot";
+      false = false;
+      float = 3.141001; # # TODO: retest on smaller floats, ex `3.14`
```
could you please add a test-line with your remianing TODO

Stunkymonkey reviewed (commented):
thanks for you contribution.

could you please add some documentation in `nixos/doc/manual/development/settings-options.section.md`

S0AndS0 commented:
> thanks for you contribution.
> 
> could you please add some documentation in `nixos/doc/manual/development/settings-options.section.md`

Hey, thanks for the review and recommendations!

I did a bit o' copy/pasting from the `toml` sub-section, with minor edits, and provided all looks good will be ready to squash commits _soon™_

S0AndS0 commented:
Commit message title updated!...  and I added a warning too, as I expect this floating-point thing to be a bit that'll cause future head-aches :-\

Tests, locally, are misbehaving on my end;

```bash
nix build --impure --reference-lock-file flake.lock ./pkgs/pkgs-lib/tests#formats.vim
#> path '/home/s0ands0/git/hub/NixOS/nixpkgs/wt/formats-add-vim/pkgs/pkgs-lib/tests' does not contain a 'flake.nix', searching up
#> error: expected flake output attribute 'legacyPackages.x86_64-linux.formats.vim' to be a derivation or path but found a
function: «lambda vim @ /nix/store/19l8nnxqrmd4q906gq647cw77dypd97i-source/pkgs/pkgs-lib/formats.nix:1031:5»
```

...  so while I updated expectations to match current conversion behaviors, I may need to rebase (and create Git Diff churn here) to speed-up testing cycle

Stunkymonkey commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 442722`
Commit: `6c986d810b0df48402a3520d71a2cccb16637838`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>tests.pkgs-lib</li>
  </ul>
</details>

h7x4 reviewed (commented):
Sorry for the late review

S0AndS0 commented on pkgs/pkgs-lib/formats.nix:
```diff
@@ -1025,6 +1027,95 @@ optionalAttrs allowAliases aliases
+                check = isVimType "raw";
+              };
+
+              vimOr = other: either other rawVim;
```
Hmm, near as I can tell my past self stole the `vimOr` bit from `elixerOr`, which also doesn't have tests x-)

The `raw = format.lib.mkRaw "{arg1, arg2 -> arg1 - arg2}"` line from `pkgs/pkgs-lib/tests/formats.nix` seems to exercise the type checking for raw, but at the moment I'm running into skill issues on if/how to add further test conditions...  I.E. I ain't ignoring ya, I just not smort enough at this moment x-)

S0AndS0 commented on pkgs/pkgs-lib/formats.nix:
```diff
@@ -1025,6 +1027,95 @@ optionalAttrs allowAliases aliases
+            // mapAttrs (_name: type: vimOr type) types;
+        };
+
+      generate = name: value: pkgs.writeText name (toVim value);
```
Tree-sitter I think is an AST generator and would require additional glue code to be a validator, or am I wrong thinking?

Also as this patch is currently intended to be generic, it does not assume it'll be used with `let b:varName = <results>` or _inventively_ injected elsewhere within Vim scripts/plugins/RC/etc files, so I'm not sure that checking results here is gonna be a winning strat.
